### ISSUE TITLE
JWT 4.0.0: Unable to set an array as "aud" claim

### ISSUE THREAD

**Issue by SvenRtbg** [lcobucci/jwt#629]
I am in the process of migrating from 3.3 via 3.4 (used right now) to 4.0.

I stumbled upon this situation: I have test cases for an OpenID-Connect library that set an array as audience claim. If I call `withClaim('aud', ['one', 'two']` I get a deprecation message. So I change the call to `permittedFor(['one','two'])` and get "Array to string conversion".

Based on the JWT RFC https://tools.ietf.org/html/rfc7519#section-4.1.3 I would say that arrays with multiple aud claim entries are allowed and possible, and I would also think they are quite common, because these test cases are there because there are some very annoying validation rules for OIDC going on depending on the "aud" claim being an array with multiple entries and an "azp" claim being present. But this is only the background info.

Bottom line: I would very much like to be able to set an "aud" claim with an array entry. Currently I cannot. Calling `permittedFor()` multiple times will overwrite the previous values.

I have also seen that the rendering of the token makes some attempts to normalize the representation of the "aud" claim. I can understand why one would do this, but unless every software consuming JWT is normalizing the difference between "single entry array" or "string", I fear this may affect situations where it is important to send either way.

**Comment by SvenRtbg** [comment]
My test method for this scenario:

    public function testSettingAudClaimArray(): void
    {
        $builder = new Builder();
        $builder->permittedFor('someone');
        $builder->permittedFor('someone2');
        $token = $builder->getToken(new None(), InMemory::plainText(''));
        $json = $token->toString();

        $parser = new Parser();
        $newToken = $parser->parse($json);
        $audience = $newToken->claims()->get('aud');
        $this->assertIsArray($audience);
        $this->assertCount(2, $audience);
    }

**Comment by lcobucci** [comment]
@SvenRtbg as you can see on the interface, the method is variadic https://github.com/lcobucci/jwt/blob/4.0.0/src/Builder.php#L20, so you can pass multiple audiences at once.

Alternatively, you can call the method more than once, as shown in your sample test scenario and covered by this test on our suite: https://github.com/lcobucci/jwt/blob/4.0.0/test/unit/Token/BuilderTest.php#L95-L118

With that said, have you tried these approaches? Is the library misbehaving in this regard or is it a documentation matter?

**Comment by SvenRtbg** [comment]
As I am moving forward right now, I'd think it is more like a documentation issue regarding the upgrade document. I understand now that looking at the latest 3.4.2 code does not really reveal how 4.0.0 looks like.

I ignored my failing tests for a moment, tried to fix all the deprecation notices, and upgraded to 4.0.0.

Let me check how things work out right now. I'm happy to give you feedback about my upgrade experience. I see some issues, but would prefer to investigate first, not causing unnecessary noise for things that turn out to be false. Thanks for your attention so far.

**Comment by lcobucci** [comment]
I see, that's good to know - at least the code is doing what it was supposed to do 🤣

I'd love to know more your upgrading experience so that we can improve the docs accordingly. If you already have some thoughts, please send us a PR.

Honestly, I'm reaaaaally terrible at documenting stuff!

**Comment by SvenRtbg** [comment]
The one thing I can see so far is that there is the gap regarding the "aud" claim transformation functionality in 3.4.2. You cannot set an array of audiences right now, and when reading the claims back, I am sure they will get "normalized" into an array with only one entry (the first one of the original claim). This will break existing code.

And there is basically nothing about this wrapping/unwrapping documented. I'd think that this would warrant additional deprecation messages in 3.4.2 and more explanaition (but I have yet to read what is already there).

**Comment by lcobucci** [comment]
Closing this then :+1:
